### PR TITLE
Fixes for EL9 support

### PR DESCRIPTION
**Pull request by jouvin**
Fixes problems detected during [24.10.0-rc4](https://github.com/quattor/release/issues/365)

### CODE REVIEW

**Comment by jouvin** [comment]
Error in `runtest` looks weird... I don't think it is related to my changes...

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 1]
```diff
@@ -113,16 +118,37 @@ variable AII_OSINSTALL_SELINUX ?= 'disabled';
...
-                + "(OS distribution location on the Quattor server)");
+    if ( !is_defined(AII_OSINSTALL_PATH) && !is_defined(BASE_OS_REPOSITORY_TEMPLATE) ) {
+        error("You need to define the variable AII_OSINSTALL_PATH or BASE_OS_REPOSITORY_TEMPLATE"
+                + " (base OS distribution location on the Quattor server)");
     };
 
     if ( match(AII_OSINSTALL_PROTOCOL, "^https?") )  {
-        "url --url " + AII_OSINSTALL_PROTOCOL + "://" + AII_OSINSTALL_SRV + AII_OSINSTALL_PATH;
+        if ( is_defined(BASE_OS_REPOSITORY_TEMPLATE) ) {
+            debug("BASE_OS_REPOSITORY_TEMPLATE=%s", to_string(BASE_OS_REPOSITORY_TEMPLATE));
```
Comment (jrha) [inline aii-ks/src/main/pan/quattor/aii/ks/config.pan:131]:
```suggestion
            debug('BASE_OS_REPOSITORY_TEMPLATE=%s', BASE_OS_REPOSITORY_TEMPLATE);
```

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 2]
```diff
@@ -354,6 +354,11 @@ variable AII_OSINSTALL_OPTION_FIREWALL ?= null;
 # default list of packages required for the initial installation
 #
 
+variable AII_OSINSTALL_VERSIONLOCK_PLUGIN = if ( OS_VERSION_PARAMS['majorversion'] >= '8' ) {
+    "python3-dnf-plugin-versionlock";
+} else {
+    "yum-plugin-versionlock";
+};
```
Comment (jrha) [inline aii-ks/src/main/pan/quattor/aii/ks/config.pan:361]: If this is OS version dependent, why don't we put it on the variant template?

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan (hunk 1 above)
Comment (jouvin) [reply aii-ks/src/main/pan/quattor/aii/ks/config.pan]: Done

**Comment by jouvin** [comment]
Rerunning the test, it seems one of the modification causes an error in one of the test. I'll investigate.

**Comment by jrha** [comment]
Don't worry too much about lint/indent on lines you haven't touched unless you have lots of spare time!

**Comment by jouvin** [comment]
@jrha trying to fix indentation issues, should not be that long...

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 3]
```diff
@@ -84,10 +89,12 @@ variable AII_OSINSTALL_PROTOCOL = value('/system/aii/osinstall/ks/osinstall_prot
...
-variable DEBUG = debug(format('%s: AII_OSINSTALL_ROOT=%s, AII_OSINSTALL_OS_VERSION=%s',
-                                OBJECT,
-                                AII_OSINSTALL_ROOT,
-                                AII_OSINSTALL_OS_VERSION));
+variable DEBUG = debug(
+    '%s: AII_OSINSTALL_ROOT=%s, AII_OSINSTALL_OS_VERSION=%s',
+    OBJECT,
+    AII_OSINSTALL_ROOT,
+    AII_OSINSTALL_OS_VERSION
+));
```
Comment (jrha) [inline aii-ks/src/main/pan/quattor/aii/ks/config.pan:97]:
```suggestion
);
```

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 4]
```diff
@@ -322,11 +351,41 @@ variable AII_OSINSTALL_OPTION_FIREWALL ?= null;
...
+#  - allow 2 types of variants : major and minor OS variants. Variants for major OS version are located in the standard configuration
+#    (as defined by AII_KS_OS_MAJOR_SPECIFIC_INCLUDE, default value should be appropriate when using QWG templates). Variants for minor
+#    OS versions are located into the related OS templates ((as defined by AII_KS_OS_MINOR_SPECIFIC_INCLUDE, default value should be appropriate when using QWG
+#    templates).  When both exist, they are both applied.
+variable  AII_KS_OS_MAJOR_SPECIFIC_INCLUDE ?=
+    if ( is_defined(OS_VERSION_PARAMS['major']) ) {
+        if_exists('quattor/aii/ks/variants/' + OS_VERSION_PARAMS['major']);
+    } else {
+        undef;
+    };
```
Comment (jrha) [inline aii-ks/src/main/pan/quattor/aii/ks/config.pan:365]:
```suggestion
variable  AII_KS_OS_MAJOR_SPECIFIC_INCLUDE ?= {
    if ( is_defined(OS_VERSION_PARAMS['major']) ) {
        if_exists('quattor/aii/ks/variants/' + OS_VERSION_PARAMS['major']);
    } else {
        undef;
    };
};
```

Comment (jrha) [reply aii-ks/src/main/pan/quattor/aii/ks/config.pan:365]: Or: 
```suggestion
variable  AII_KS_OS_MAJOR_SPECIFIC_INCLUDE ?= if ( is_defined(OS_VERSION_PARAMS['major']) ) {
    if_exists('quattor/aii/ks/variants/' + OS_VERSION_PARAMS['major']);
} else {
    undef;
};
```

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 5]
```diff
@@ -322,11 +351,41 @@ variable AII_OSINSTALL_OPTION_FIREWALL ?= null;
...
+    debug('KS specific configuration for OS major version: ' + to_string(AII_KS_OS_MAJOR_SPECIFIC_INCLUDE));
+    AII_KS_OS_MAJOR_SPECIFIC_INCLUDE;
+};
+# Existence of OS_VERSION_PARAMS['minor'] is used as a QWG signature
+variable  AII_KS_OS_MINOR_SPECIFIC_INCLUDE ?=
+    if ( is_defined(OS_VERSION_PARAMS['minor']) ) {
+        if_exists('config/quattor/ks');
+    } else {
+        undef;
+};
```
Comment (jrha) [inline aii-ks/src/main/pan/quattor/aii/ks/config.pan:376]: Same here:
```suggestion
variable  AII_KS_OS_MINOR_SPECIFIC_INCLUDE ?= {
    if ( is_defined(OS_VERSION_PARAMS['minor']) ) {
        if_exists('config/quattor/ks');
    } else {
        undef;
    };
};
```

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan (hunk 3 above)
Comment (jouvin) [reply aii-ks/src/main/pan/quattor/aii/ks/config.pan:99]: Fixed

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan (hunk 4 above)
Comment (jouvin) [reply aii-ks/src/main/pan/quattor/aii/ks/config.pan]: In fact I removed them at the same time you commenting on it.

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan (hunk 5 above)
Comment (jouvin) [reply aii-ks/src/main/pan/quattor/aii/ks/config.pan]: Fixed

**Comment by jouvin** [comment]
@jrha should be ok now, except that the tests are failing in `aii-ks` that complains that `debug` should have only one argument... which is no longer true with the implicit call to `format()`. Not sure where it is coming from...

**Comment by jrha** [comment]
Oh, it's using panc 10.2…?
```
[INFO] --- panc-maven-plugin:10.2:check-syntax (check-generated-pan-syntax) @ ks ---
syntax error [/__w/aii/aii/aii-ks/target/pan/quattor/aii/ks/config.pan:108.18-113.1]
```

**Comment by jrha** [comment]
https://github.com/quattor/aii/pull/353

**Comment by jrha** [comment]
Thanks to @wpoely86 for merging all the build tools updates!
You should be able to just rebase this PR now.

**Comment by jouvin** [comment]
Looks good now!

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 6]
```diff
@@ -65,16 +67,21 @@ variable AII_OSINSTALL_PATH ?= undef;
 # defaults to http
 #
 variable AII_OSINSTALL_PROTOCOL ?=
-    if ( exists("/system/aii/osinstall/ks/osinstall_protocol") ) {
-        error('Use AII_OSINSTALL_PROTOCOL to define installation protocol');
-    } else {
-        'http';
-    };
+if ( exists("/system/aii/osinstall/ks/osinstall_protocol") ) {
```
Comment (jrha) [inline aii-ks/src/main/pan/quattor/aii/ks/config.pan:70]:
```suggestion
variable AII_OSINSTALL_PROTOCOL ?= if (exists("/system/aii/osinstall/ks/osinstall_protocol")) {
```

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 7]
```diff
@@ -84,25 +91,27 @@ variable AII_OSINSTALL_PROTOCOL = value('/system/aii/osinstall/ks/osinstall_prot
...
+    OBJECT,
+    AII_OSINSTALL_ROOT,
+    AII_OSINSTALL_OS_VERSION
+);
 
 # AII_OSINSTALL_SUBURL allows to specify a sub-url under root/version
 # (e.g. /base)
 
 variable AII_OSINSTALL_SUBURL ?= undef;
 variable AII_OSINSTALL_PATH ?=
```
Comment (jrha) [inline aii-ks/src/main/pan/quattor/aii/ks/config.pan:105]:
```suggestion
```

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 8]
```diff
@@ -324,11 +359,35 @@ variable AII_OSINSTALL_OPTION_FIREWALL ?= null;
...
+#  - including this at the end allow to undefine tree elements, and remain compatible with other (previous) OSes
+#  - allow 2 types of variants : major and minor OS variants. Variants for major OS version are located in the standard configuration
+#    (as defined by AII_KS_OS_MAJOR_SPECIFIC_INCLUDE, default value should be appropriate when using QWG templates). Variants for minor
+#    OS versions are located into the related OS templates ((as defined by AII_KS_OS_MINOR_SPECIFIC_INCLUDE, default value should be appropriate when using QWG
+#    templates).  When both exist, they are both applied.
+variable  AII_KS_OS_MAJOR_SPECIFIC_INCLUDE ?= if ( is_defined(OS_VERSION_PARAMS['major']) ) {
+    if_exists('quattor/aii/ks/variants/' + OS_VERSION_PARAMS['major']);
+};
+include {
+    debug('KS specific configuration for OS major version: ' + to_string(AII_KS_OS_MAJOR_SPECIFIC_INCLUDE));
```
Comment (jrha) [inline aii-ks/src/main/pan/quattor/aii/ks/config.pan:372]:
```suggestion
    debug('KS specific configuration for OS major version: %s', AII_KS_OS_MAJOR_SPECIFIC_INCLUDE);
```

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 9]
```diff
@@ -324,11 +359,35 @@ variable AII_OSINSTALL_OPTION_FIREWALL ?= null;
...
+    AII_KS_OS_MAJOR_SPECIFIC_INCLUDE;
+};
+# Existence of OS_VERSION_PARAMS['minor'] is used as a QWG signature
+variable  AII_KS_OS_MINOR_SPECIFIC_INCLUDE ?= if ( is_defined(OS_VERSION_PARAMS['minor']) ) {
+    if_exists('config/quattor/ks');
+};
+include {
+    debug('KS specific configuration for OS minor release: ' + to_string(AII_KS_OS_MINOR_SPECIFIC_INCLUDE));
```
Comment (jrha) [inline aii-ks/src/main/pan/quattor/aii/ks/config.pan:380]:
```suggestion
    debug('KS specific configuration for OS minor release: %s', AII_KS_OS_MINOR_SPECIFIC_INCLUDE);
```

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 10]
```diff
@@ -389,54 +450,29 @@ variable AII_OSINSTALL_PACKAGES ?= list(
...
 variable AII_PROFILE_PATH ?= "/profiles";
 variable AII_OSINSTALL_NODEPROFILE ?=
-    if (AII_USE_CCM) {
-        if (exists("/software/components/ccm/profile") && !(value("/software/components/ccm/profile") == '' )) {
-            value("/software/components/ccm/profile");
-        } else {
-            error("Can't find value for the profile url at /software/components/ccm/profile. " +
-                    "If you don't use ccm, set the variable AII_USE_CCM to false.");
-        };
+if (AII_USE_CCM) {
```
Comment (jrha) [inline aii-ks/src/main/pan/quattor/aii/ks/config.pan:463]:
```suggestion
variable AII_OSINSTALL_NODEPROFILE ?= if (AII_USE_CCM) {
```

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 11]
```diff
@@ -389,54 +450,29 @@ variable AII_OSINSTALL_PACKAGES ?= list(
...
+variable AII_ACK_LIST ?= list(format("%s://%s%s", AII_ACK_PROTOCOL, AII_ACK_SRV, AII_ACK_CGI));
 "acklist" ?= AII_ACK_LIST;
 
 #
 # Set the location of the node profile
 #
+include 'components/ccm/config';
 variable AII_USE_CCM ?= exists("/software/components/ccm") && is_defined("/software/components/ccm");
 variable AII_PROFILE_PATH ?= "/profiles";
 variable AII_OSINSTALL_NODEPROFILE ?=
```
Comment (jrha) [inline aii-ks/src/main/pan/quattor/aii/ks/config.pan:462]:
```suggestion
```

File: aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan [hunk 12]
```diff
@@ -113,20 +113,16 @@ variable AII_NBP_ROOT ?= {
 };
 
 "/system/aii/nbp/pxelinux/kernel" ?=
-    if ( is_defined(AII_OSINSTALL_OS_VERSION) ) {
-        format('%s/vmlinuz', AII_NBP_ROOT);
-    } else {
-        undef;
-    };
+if ( is_defined(AII_NBP_ROOT) ) {
```
Comment (jrha) [inline aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan:116]:
```suggestion
"/system/aii/nbp/pxelinux/kernel" ?= if ( is_defined(AII_NBP_ROOT) ) {
```

File: aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan [hunk 13]
```diff
@@ -99,7 +99,7 @@ variable AII_NBP_ROOT ?= {
     if (LPXELINUX) {
         root = LPXELINUX_ROOT;
     } else {
-        toks =  matches(AII_OSINSTALL_OS_VERSION, '^(\w+?)[_\-](.*)');
+        toks =  matches(AII_OSINSTALL_OS_VERSION, '^([\w\.]+?)[_\-](.*)$');
```
Comment (jrha) [inline aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan:102]:
```suggestion
        toks = matches(AII_OSINSTALL_OS_VERSION, '^([\w\.]+?)[_\-](.*)$');
```

File: aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan [hunk 14]
```diff
@@ -113,20 +113,16 @@ variable AII_NBP_ROOT ?= {
...
-    } else {
-        undef;
-    };
+if ( is_defined(AII_NBP_ROOT) ) {
+    format('%s/vmlinuz', AII_NBP_ROOT);
+};
 
 variable AII_NBP_INITRD ?= "initrd.img";
 
 "/system/aii/nbp/pxelinux/initrd" ?=
```
Comment (jrha) [inline aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan:122]:
```suggestion
```

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 15]
```diff
@@ -65,16 +67,21 @@ variable AII_OSINSTALL_PATH ?= undef;
 # defaults to http
 #
 variable AII_OSINSTALL_PROTOCOL ?=
```
[the remark this answers is not in the document]
Comment (jouvin) [reply aii-ks/src/main/pan/quattor/aii/ks/config.pan:69]: Done

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan (hunk 6 above)
Comment (jouvin) [reply aii-ks/src/main/pan/quattor/aii/ks/config.pan:70]: Done

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan (hunk 7 above)
Comment (jouvin) [reply aii-ks/src/main/pan/quattor/aii/ks/config.pan:105]: Done

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan [hunk 16]
```diff
@@ -84,25 +91,27 @@ variable AII_OSINSTALL_PROTOCOL = value('/system/aii/osinstall/ks/osinstall_prot
...
 variable AII_OSINSTALL_PATH ?=
-    if ( is_defined(AII_OSINSTALL_ROOT) && is_defined(AII_OSINSTALL_OS_VERSION) ) {
-        path = AII_OSINSTALL_ROOT + '/' + AII_OSINSTALL_OS_VERSION;
-        if ( is_defined(AII_OSINSTALL_SUBURL) ) {
-            path = path + AII_OSINSTALL_SUBURL;
-        };
-        path;
-    } else {
-        debug('AII_OSINSTALL_ROOT or AII_OSINSTALL_OS_VERSION undefined: cannot define AII_OSINSTALL_PATH');
+if ( is_defined(AII_OSINSTALL_ROOT) && is_defined(AII_OSINSTALL_OS_VERSION) ) {
```
[the remark this answers is not in the document]
Comment (jouvin) [reply aii-ks/src/main/pan/quattor/aii/ks/config.pan:106]: Done

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan (hunk 8 above)
Comment (jouvin) [reply aii-ks/src/main/pan/quattor/aii/ks/config.pan:372]: Done

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan (hunk 9 above)
Comment (jouvin) [reply aii-ks/src/main/pan/quattor/aii/ks/config.pan:380]: Done

File: aii-ks/src/main/pan/quattor/aii/ks/config.pan (hunk 10 above)
Comment (jouvin) [reply aii-ks/src/main/pan/quattor/aii/ks/config.pan:463]: Done

File: aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan (hunk 12 above)
Comment (jouvin) [reply aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan:116]: Done

File: aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan [hunk 17]
```diff
@@ -113,20 +113,16 @@ variable AII_NBP_ROOT ?= {
 };
 
 "/system/aii/nbp/pxelinux/kernel" ?=
```
[the remark this answers is not in the document]
Comment (jouvin) [reply aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan:115]: Done

File: aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan (hunk 13 above)
Comment (jouvin) [reply aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan:102]: Done

Comment (jouvin) [reply aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan:102]: Strange if was not found by `panlint`...

File: aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan (hunk 14 above)
Comment (jouvin) [reply aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan:122]: Done

File: aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan [hunk 18]
```diff
@@ -113,20 +113,16 @@ variable AII_NBP_ROOT ?= {
...
 
 variable AII_NBP_INITRD ?= "initrd.img";
 
 "/system/aii/nbp/pxelinux/initrd" ?=
-    if ( is_defined(AII_OSINSTALL_OS_VERSION) ) {
-        format('%s/%s', AII_NBP_ROOT, AII_NBP_INITRD);
-    } else {
-        undef;
-    };
+if ( is_defined(AII_NBP_ROOT) ) {
```
[the remark this answers is not in the document]
Comment (jouvin) [reply aii-pxelinux/src/main/pan/quattor/aii/pxelinux/config.pan]: Done

**Review by jrha — APPROVED** [review]

**Pull request merged**